src/bblgum/core.clj

```diff
@@ -2,24 +2,43 @@
   (:require [bblgum.impl :as i]))
 
 (defn gum
-  "Main driver fn for using gum: https://github.com/charmbracelet/gum
+  "Main driver of bblgum.
 
-  Options map:
-  cmd: The interaction command. Can be a keyword. Required
-  opts: A map of options to be passed as optional params to gum
-  args: A vec of args to be passed as positional params to gum
-  in: An input stream than can be passed to gum
-  as: Coerce the output. Currently supports :bool, :ignored or defaults to a seq of strings
-  gum-path: Path to the gum binary. Defaults to gum
+  You can call `gum` like this:
+
+  (gum :command [\"args vector\"] :opt \"value\" :opt2 \"value2\")
+
+  There are several special opts, that are handled by the library:
+  :in - An input stream than can be passed to gum
+  :as - Coerce the output. Currently supports :bool, :ignored or defaults to a seq of strings
+  :gum-path - Path to the gum binary. Defaults to gum
+
+  All other opts are passed to the `gum` CLI. Consult `gum CMD --help` to see available options.
+  To pass flags like `--directory` use `:directory true`. Always use full names of the options.
+
+  Usage examples
+  --------------
+  Command only:
+  (gum :file)
+
+  Command with args:
+  (gum :choose [\"arg\" \"arg2\"])
+
+  Command with opts:
+  (gum :file :directory true)
+
+  Command with arguments and options:
+  (gum :choose [\"arg1\" \"arg2\"] :header \"Choose an option\")
 
   Returns a map of:
   status: The exit code from gum
   result: The output from the execution: seq of lines or coerced via :as."
-  [{:keys [cmd opts args in as gum-path]}]
-  (when-not cmd
-    (throw (IllegalArgumentException. ":cmd must be provided or non-nil")))
-
-  (let [gum-path (or gum-path "gum")
+  [cmd & [args & opts]]
+  (when-not
+      (or (keyword? cmd)
```

Comment:
wouldnt this fail with the old direct map invocation then? `(gum {...})`

Reply:
true

Reply:
i was okay with the multi arity one though fyi 😛

Reply:
no hurries, thanks a lot for all the effort though anyways!

Reply:
Well, I tried this:

```clojure
(defn invalid-args? [cmd & [args & opts]]
  (cond (and (map? cmd) (not (and (empty? args) (empty? opts)))) "if you provide an entire action map as first argument, other arguments are not allowed"
        (map? cmd) nil
        (not (or (keyword? cmd) (string? cmd))) "cmd must be a keyword or a string / alternatively it can be an entire options map"
        :else nil))

(comment
  (invalid-args? nil) ;; => "cmd must be a keyword or a string / alternatively it can be an entire options map"
  (invalid-args? {})  ;; => nil
  (invalid-args? 1)   ;; => "cmd must be a keyword or a string / alternatively it can be an entire options map"
  (invalid-args? :cmd :opt :val)                                                ;; => nil
  (invalid-args? :cmd ["arg"] :opt :val :in "in")                               ;; => nil
  (invalid-args? :cmd :opt1 :val1 :opt2 :val2 :in "in" :as :bool)               ;; => nil
  (invalid-args? :cmd ["arg" "arg"] :opt1 :val1 :opt2 :val2 :in "in" :as :bool) ;; => nil
  (invalid-args? {} ["foo"] :as :bool) ;; => "if you provide an entire action map as first argument, other arguments are not allowed"
  (invalid-args? {} :foo :bar)
  )
```

And it handles everything fine, except the last case in the comment — this throws an exception by itself. I don't know if that's fine, but if we keep this approach, we'll have this in gum:

```clojure
  (when-let [err (apply invalid-args? cmd args opts)]
    (throw (IllegalArgumentException. err)))
```

Reply:
So there's an edge case when user calls `gum` with a map as a first arg, and then a keyword, which breaks the destructuring itself.

Reply:
It wouldn't break multi-arity version though, so I now think that the multi-arity version was actually superior. Will bring it back and add some validation to it.

Reply:
re: invalid-args?

I feel the multi arity is cleaner, signalling the intent better.

> So there's an edge case when user calls gum with a map as a first arg, and then a keyword, which breaks the destructuring itself.

Its fine i think. LSP should error there i think? It's either that error or the `IllegalArgumentException`.